browser-extension/mv2/src/client/js/scriptRuleHandler.js

```diff
@@ -128,24 +128,105 @@ RQ.ScriptRuleHandler.includeJSScriptsInOrder = function (scripts, callback, inde
+      });
+      RQ.ScriptRuleHandler.embedBlocks("script", scriptBlocks);
+    }
   }
-
   typeof callback === "function" && callback();
 };
 
 RQ.ScriptRuleHandler.includeCSS = function (script, callback) {
   if (script.type === RQ.SCRIPT_TYPES.URL) {
+    if (script.wrapperElement) {
+      const styleLinkBlock = RQ.ScriptRuleHandler.parseSpecificHTMLStringBlocks(script.wrapperElement, "link")?.[0];
+      if (styleLinkBlock) {
+        styleLinkBlock.attributes["href"] = script.value;
+        styleLinkBlock.innerText = "";
+        RQ.ScriptRuleHandler.embedBlocks("link", [styleLinkBlock]);
+        return;
+      }
+    }
     RQ.ClientUtils.addRemoteCSS(script.value);
   } else if (script.type === RQ.SCRIPT_TYPES.CODE) {
-    RQ.ClientUtils.embedCSS(script.value);
+    const styleBlocks = RQ.ScriptRuleHandler.parseSpecificHTMLStringBlocks(script.value, "style");
+    if (!styleBlocks || !styleBlocks.length) {
+      RQ.ClientUtils.embedCSS(script.value);
+    } else {
+      RQ.ScriptRuleHandler.embedBlocks("style", styleBlocks);
+    }
   }
 
   typeof callback === "function" && callback();
 };
+
+RQ.ScriptRuleHandler.parseSpecificHTMLStringBlocks = function (rawCode, htmlElementName) {
+  if (!htmlElementName || htmlElementName === "body") throw new Error("htmlElementName is empty");
+  const parser = new DOMParser();
+  // if it is a normal code block, string will be added to body, hence not parsing body
+  const doc = parser.parseFromString(rawCode, "text/html");
+  const blocks = Array.from(doc.getElementsByTagName(htmlElementName));
+  const blockDetails = blocks.map((htmlBlock, index) => {
+    return {
+      index: index,
+      innerText: htmlBlock.innerHTML.trim(),
+      attributes: Array.from(htmlBlock.attributes).reduce((attrs, attr) => {
+        attrs[attr.name] = attr.value ?? "";
+        return attrs;
+      }, {}),
+    };
+  });
+  return blockDetails;
+};
+
+RQ.ScriptRuleHandler.embedBlocks = function (htmlElementName, blockDetails) {
+  if (!htmlElementName || htmlElementName === "body") throw new Error("htmlElementName is empty");
+  blockDetails.forEach((block) => {
+    const htmlBlock = document.createElement(htmlElementName);
+    htmlBlock.className = RQ.ClientUtils.getScriptClassAttribute();
+
+    htmlBlock.appendChild(document.createTextNode(block.innerText));
+
+    Object.keys(block.attributes).forEach((attr) => {
+      if (attr === "class") {
+        // to also add the RQ specific class name
+        htmlBlock.className += RQ.ClientUtils.getScriptClassAttribute() + block.attributes[attr];
+      }
+      htmlBlock.setAttribute(attr, block.attributes[attr]);
+    });
+
+    RQ.ScriptRuleHandler.appendToDom(htmlBlock);
+  });
+};
+
+RQ.ScriptRuleHandler.appendToDom = function (htmlNode) {
+  const parent = document.head || document.documentElement;
+  try {
+    parent.appendChild(htmlNode);
+  } catch (error) {
+    console.error(error);
```

Comment:
In case of error, shouldn't we log something user friendly instead of the actual error?